vowpalwabbit/core/src/confidence_sequence_robust.cc

```diff
@@ -0,0 +1,403 @@
+    : log_xi(std::log1p(xi - 1))
+    , log_xi_m1(std::log1p(xi - 2.0))
+    , lambda_max(lambda_max)
+    , zeta_r(
+#if !defined(__APPLE__) && !defined(_WIN32)
+          std::riemann_zeta(r)
+#else
+          1.6449340668482264  // std::riemann_zeta(r) -- Assuming r=2.0 is constant
+#endif
+              )
+    , scale_fac(0.5 * (1.0 + polylog(r, eta) / (eta * zeta_r)))
+    , log_scale_fac(std::log1p(scale_fac - 1.0))
+    , t(0)
+    , gt(k)
+{
+  assert(0.0 < eta && eta < 1.0);
+  assert(r > 1.0);
+  assert(0.0 < scale_fac && scale_fac < 1.0);
+  assert(0.0 < lambda_max && lambda_max <= 1.0 + (-0.15859433956303937));  // sc.lambertw(-exp(-2)) in Python
+  assert(1.0 < xi);
+}
+
+double countable_discrete_base::get_ci(double alpha) const { return lb_log_wealth(alpha); }
+
+double countable_discrete_base::get_lam_sqrt_tp1(double j) const
+{
+  double log_den = (j + 0.5) * log_xi - 0.5 * std::log(t + 1);
+  return lambda_max * std::exp(-log_den);
+}
+
+double countable_discrete_base::get_v_impl(std::map<uint64_t, double>& memo, uint64_t j) const
+{
+  if (memo.find(j) == memo.end()) { memo[j] = get_v(get_lam_sqrt_tp1(j)); }
+  return memo[j];
+}
+
+double countable_discrete_base::log_sum_exp(const std::vector<double>& a) const
+{
+  double max_exp = *std::max_element(a.begin(), a.end());
+  double ret = 0.0;
+  for (const double& v : a) { ret += std::exp(v - max_exp); }
+
+  return std::log(ret) + max_exp;
+}
+
+double countable_discrete_base::log_wealth_mix(
+    double mu, double s, double thres, std::map<uint64_t, double>& memo) const
+{
+  double sqrt_tp1 = std::sqrt(t + 1);
+  double y = s / sqrt_tp1 - (t / sqrt_tp1) * mu;
+  std::vector<double> log_es;
+  std::vector<double> log_ws;
+  for (const std::pair<const unsigned long, double>& memo_p : memo)
+  {
+    uint64_t j = memo_p.first;
+    double v = memo_p.second;
+    log_es.push_back(get_lam_sqrt_tp1(j) * y - v);
+    log_ws.push_back(get_log_weight(j));
+  }
+
+  uint64_t j = log_es.size();
+  while (true)
+  {
+    std::vector<double> combined;
+    for (size_t i = 0; i < log_es.size(); ++i) { combined.push_back(log_es[i] + log_ws[i]); }
+    double lower_bound = log_sum_exp(combined);
+
+    if (lower_bound >= thres) { return lower_bound; }
+
+    // quasiconcave after the maximum, otherwise lower bound variance with 0
+    double max_of_begin = std::numeric_limits<int>::min();
+    for (size_t i = 0; i < log_es.size() - 1; ++i) { max_of_begin = std::max(max_of_begin, log_es[i]); }
+    double log_upper_exp = (log_es.back() < max_of_begin) ? log_es.back() : get_lam_sqrt_tp1(j) * y;
+    double upper_bound = log_sum_exp(std::vector<double>({lower_bound, get_log_remaining_weight(j) + log_upper_exp}));
+
+    if (upper_bound < thres) { return upper_bound; }
+
+    double v = get_v_impl(memo, j);
+    log_es.push_back(get_lam_sqrt_tp1(j) * y - v);
+    log_ws.push_back(get_log_weight(j));
+    j += 1;
+  }
+}
+
+double countable_discrete_base::root_brentq(
```

Comment:
adding a unit test for root_brentq might be a good idea, just like the simple one you had and shared before

Reply:
added check